### PR TITLE
Refactor contract content into types, expand testing, add a lot of Uniswap constructs

### PR DESCRIPTION
**Pull request by gravaton**
This PR includes a big pile of things

- Updated constructs for use when dealing with Uniswap pools.  These new constructs should let us operate on pools in a much more ergonomic manner when we have to calculate or simulate things
- Proper code to produce contract-compatible output from the results of our matching algo and ToB order, now refactored to be pulled properly back into the type crate and operate as simple conversions from underlying data types.
- Appropriate types for such constructs and contract bindings that can be used to deploy and communicate with our contracts.  Additionally a small script to generate/regenerate contract bindings
- Easy to stand up Alloy environments with prepackaged deployment of our contract base for further testing
- Improvements in order generators with more to come
- Redid the AMM types and constructs to both be better named and easier to test
- Fixed a small PADE bug when decoding generics
- Shrank a few modules by moving things into types where it will have a better home

### CODE REVIEW

File: bin/contract_test/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                    >
+                >
+            >
+        >,
+        alloy::providers::fillers::WalletFiller<EthereumWallet>
+    >,
+    alloy::providers::RootProvider<alloy::pubsub::PubSubFrontend>,
+    alloy::pubsub::PubSubFrontend,
+    Ethereum
+>;
```
[the remark this answers is not in the document]
Comment (gravaton) [reply bin/contract_test/src/lib.rs:30]: Hah yeah it's really annoying that these provider types get so huge!

File: bin/contract_test/src/contract_payloads.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    #[derive(PadeEncode, Debug)]
+    struct Pair {
+        uint16 index0;
+        uint16 index1;
+        uint16 store_index;
+        uint256 price_1over0;
+    }
+}
```
Comment (Will-Smith11) [inline bin/contract_test/src/contract_payloads.rs:22]: why do we not use the ones defined in types here?

File: crates/types/src/contract_payloads/rewards.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    pub swap_in_quantity: u128,
+    pub rewards_update:   RewardsUpdate
+}
+
+#[derive(PadeEncode, Debug)]
+pub struct MockContractMessage {
+    pub assets: Vec<Asset>,
+    pub pairs:  Vec<Pair>,
+    pub update: PoolUpdate
+}
```
Comment (Will-Smith11) [inline crates/types/src/contract_payloads/rewards.rs:25]: isn't there duplicates of this in `bin/contract_test`?

File: crates/types/src/matching/ray.rs [hunk 4]
```diff
@@ -143,19 +158,19 @@ impl Ray {
         self.into()
     }
 
-    /// Calculates a price ratio t0/t1
+    /// Calculates a price ratio t1/t0
     pub fn calc_price(t0: U256, t1: U256) -> Self {
```
Comment (Will-Smith11) [inline crates/types/src/matching/ray.rs:162]: good to know ray math exists for my stuff

File: testing-tools/src/type_generator/orders/mod.rs [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+    pub fn asset_in(self, asset_in: Address) -> Self {
+        Self { asset_in, ..self }
+    }
+
+    pub fn asset_out(self, asset_out: Address) -> Self {
+        Self { asset_out, ..self }
+    }
+
+    pub fn amount(self, amount: u128) -> Self {
+        Self { amount, ..self }
```
Comment (Will-Smith11) [inline testing-tools/src/type_generator/orders/mod.rs:93]: cool didn't know you could do this

**Review by Will-Smith11 — APPROVED** [review]
lgtm

File: bin/contract_test/src/contract_payloads.rs (hunk 2 above)
Comment (gravaton) [reply bin/contract_test/src/contract_payloads.rs:22]: Ah, can do, this was a very quick-and-dirty test bin I wrote that isn't expected to remain so this was some scratch work I was messing with

File: crates/types/src/contract_payloads/rewards.rs (hunk 3 above)
Comment (gravaton) [reply crates/types/src/contract_payloads/rewards.rs:25]: Yep, removed

**Pull request merged**